compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -408,16 +408,29 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
                 } else {
                     ty::WithOptConstParam::unknown(closure_id)
                 };
-                let (closure_thir, expr) = self.tcx.thir_body(closure_def).unwrap_or_else(|_| {
-                    (self.tcx.alloc_steal_thir(Thir::new()), ExprId::from_u32(0))
-                });
-                let closure_thir = &closure_thir.borrow();
-                let hir_context = self.tcx.hir().local_def_id_to_hir_id(closure_id);
-                let mut closure_visitor =
-                    UnsafetyVisitor { thir: closure_thir, hir_context, ..*self };
-                closure_visitor.visit_expr(&closure_thir[expr]);
-                // Unsafe blocks can be used in closures, make sure to take it into account
-                self.safety_context = closure_visitor.safety_context;
+                if let Ok((closure_thir, expr)) = self.tcx.thir_body(closure_def) {
+                    let closure_thir = &closure_thir.borrow();
+                    let hir_context = self.tcx.hir().local_def_id_to_hir_id(closure_id);
+                    let mut closure_visitor =
+                        UnsafetyVisitor { thir: closure_thir, hir_context, ..*self };
+                    closure_visitor.visit_expr(&closure_thir[expr]);
+                    // Unsafe blocks can be used in closures, make sure to take it into account
+                    self.safety_context = closure_visitor.safety_context;
+                }
```

Comment:
could this be a method called for closures and const blocks instead of having (almost?) the same code twice?

Reply:
Done